File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/schema/read/AttSchemaUtils.java
package ru.citeck.ecos.records3.record.op.atts.schema.read;

import ecos.com.fasterxml.jackson210.databind.JsonNode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import ru.citeck.ecos.records3.record.op.atts.schema.SchemaAtt;

import java.util.*;

public class AttSchemaUtils {

    @Nullable
    public static Class<?> getFlatScalarType(@NotNull SchemaAtt att) {

        String name = att.getName();

        if (att.isScalar()) {
            switch (name) {
                case "json":
                    return JsonNode.class;
                case "bool":
                    return Boolean.class;
                case "num":
                    return Double.class;
                default:
                    return String.class;
            }
        }

        List<SchemaAtt> inner = att.getInner();

        if (inner.size() == 0) {
            throw new AttReadException(att.getAlias(), att.getName(),
                "Attribute is not a scalar and doesn't have inner attributes");
        } else if (inner.size() == 1) {
            return getFlatScalarType(inner.get(0));
        }

        return null;
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/atts/schema/resolver/AttSchemaUtils.kt
package ru.citeck.ecos.records3.record.atts.schema.resolver

import ru.citeck.ecos.records3.record.atts.schema.SchemaAtt
import java.util.ArrayList

object AttSchemaUtils {

    /**
     * Replace aliases by name and join attributes.
     */
    fun simplifySchema(schema: List<SchemaAtt>): List<SchemaAtt> {
        if (schema.isEmpty()) {
            return schema
        }
        val result = LinkedHashMap<String, SchemaAtt.Builder>()
        for (att in schema) {
            var resAtt = result[att.name]
            if (resAtt == null) {
                resAtt = att.copy()
                    .withAlias(att.name)
                    .withProcessors(emptyList())
                result[att.name] = resAtt
            } else {
                resAtt.withMultiple(resAtt.multiple || att.multiple)
                val innerAtts = ArrayList<SchemaAtt>()
                innerAtts.addAll(resAtt.inner)
                innerAtts.addAll(att.inner)
                resAtt.withInner(innerAtts)
            }
        }
        val resultAtts = ArrayList<SchemaAtt>()
        for (att in result.values) {
            att.withInner(simplifySchema(att.inner))
            resultAtts.add(att.build())
        }
        return resultAtts
    }
}
